modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,660 @@
+
+        /** Stop file writer future. */
+        GridFutureAdapter<Void> stopFut = new GridFutureAdapter<>();
+
+        /** Cached strings by id. */
+        private final ConcurrentHashMap<String, Short> stringIds = new ConcurrentHashMap<>();
+
+        /** String id generator. */
+        private final AtomicInteger idsGen = new AtomicInteger();
+
+        /** {@code True} if the small buffer warning message logged. */
+        private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+        /** {@code True} if worker stopped due to maximum file size reached. */
+        private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+        /**
+         * @param ctx Kernal context.
+         * @param fileIo Performance statistics file I/O.
+         * @param maxFileSize Maximum file size in bytes.
+         * @param bufferSize Off heap buffer size in bytes.
+         * @param flushBatchSize Minimal batch size to flush in bytes.
+         * @param log Logger.
+         */
+        FileWriter(GridKernalContext ctx, FileIO fileIo, long maxFileSize, int bufferSize, int flushBatchSize,
+            IgniteLogger log) {
+            super(ctx.igniteInstanceName(), "performance-statistics-writer%" + ctx.igniteInstanceName(), log);
+
+            this.fileIo = fileIo;
+            this.flushBatchSize = flushBatchSize;
+
+            ringByteBuffer = new SegmentedRingByteBuffer(bufferSize, maxFileSize, BufferMode.DIRECT);
+
+            ringByteBuffer.init(0);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+            while (!isCancelled() && !Thread.interrupted()) {
+                blockingSectionBegin();
+
+                try {
+                    synchronized (this) {
+                        while (readyForFlushSize.get() < flushBatchSize && !isCancelled())
+                            wait();
+                    }
+
+                    flushBuffer();
+                }
+                finally {
+                    blockingSectionEnd();
+                }
+            }
+
+            fileWriter = null;
+
+            ringByteBuffer.close();
+
+            // Make sure that all producers released their buffers to safe deallocate memory.
+            ringByteBuffer.poll();
+
+            ringByteBuffer.free();
+
+            U.closeQuiet(fileIo);
+
+            stringIds.clear();
+
+            stopFut.onDone();
+
+            log.info("Performance statistics writer stopped.");
+        }
+
+        /** @return Unique per file string identifier. {@code Null} if there is no cached identifier. */
+        Short stringId(String str) {
+            return stringIds.get(str);
+        }
+
+        /** @return Generate unique per file string identifier. {@code -1} if max cached limit exceeded. */
+        short generateStringId(String str) {
+            if (idsGen.get() > MAX_CACHED_STRING_COUNT)
```

Comment:
Can we use the string hash code as an identifier?
Storing `ConcurrentHashMap` with the query string will create GC pressure if the user executes many different queries.

We can use Set<Integer> or even BloomFilter implementation to check if some string already seen in the current statistics.

Reply:
I want to avoid collisions. It may lead to wrong results interpretation.
`MAX_CACHED_STRING_COUNT` can be decreased to reduce GC pressure.

Reply:
What collisions do you have in mind? A collision of hash codes of the two strings?

Reply:
I have implemented with `Set` of hashcodes